### ISSUE TITLE
Proper Device Profiles Implementation

### ISSUE THREAD

**Issue by Stetsed** [streamyfin/streamyfin#191]
This Issue is to track progress on a proper device profile implementation discussed in the dev chat to allow less needed transcoding and for it to not suck.

Jellyfin API Endpoint:
https://api.jellyfin.org/#tag/MediaInfo/operation/GetPostedPlaybackInfo

Jellyfin API takes in "DeviceProfile", which then contains the diffrent profiles, an example POST request for this is
![image](https://github.com/user-attachments/assets/ae3adc8b-b10a-40d7-8fa4-e7931bab82bb)

After the switch to VLC has been made containers won't be of much concern due to VLC having the container support. However due to the wide range of devices we do need a way to find what codec support exists on the device, on iOS this is relativley easy:

To check an individual codec:
https://developer.apple.com/documentation/videotoolbox/vtishardwaredecodesupported(_:)

Which takes an argument from the list of codecs apple provides:
https://developer.apple.com/documentation/coremedia/cmvideocodectype

Which would allow for this pseudo code, which also makes it easy to expand to more codecs as long as apple adds support for them.
```rust
fn check_codecs() -> supportedCodecs {
    // Codecs that need to be checked
    let codecs_to_check = [
        kCMVideoCodecType_H264,
        kCMVideoCodecType_H265,
        kCMVideoCodecType_AV1,
    ];

    // a struct that contains the codecs we want
    let supportedCodecs = {h264: false, h265, false, av1, false};

    let supporteCodecs 
    for codec in codecs_to_check {
        if VTIsHardwareDecodeSupported(codec) {
            // Might need to parse here, as it gives the raw codec type, not the usually accepted
            // representation such as h264, h265 etc.
            supportedCodecs.codec = true;
        }
    }
    return supportedCodecs
}
```

For android something similar could be used, but it seems like naming for the decoders isn't standardized so you would probally wanna use some string search(possibly in the form of regex to make it less sensitive to manufacters bullshit). Below post was what I found and seems to give what we wanted and should allow with integration like the above but again, possibly with more parsing.

https://stackoverflow.com/questions/11791473/get-supported-codec-for-android-device#26589396

And then with the supported codecs we could hook this into the existing post requests, but instead of using the current static profiles(e.g [ios](https://github.com/fredrikburmester/streamyfin/blob/bbd12c540ad2cb9085f228be164479ac59788039/utils/profiles/ios.js). We would use our own with the codecs we have gotten above to form them.

**Comment by fredrikburmester** [comment]
VLC supports all codecs using software decoding (if not supported using hw), so direct play will always be possible if we want.

**Comment by Stetsed** [comment]
> VLC supports all codecs using software decoding (if not supported using hw), so direct play will always be possible if we want.

Correct, but on mobile devices we do not want to use software decoding, it's more about container support cuz it means we can direct play with any hardware supported codec and not have to care about the container

**Comment by neingeist** [comment]
See also #139. There are also special cases like HEVC being supported in HW, but not with 10 bit color depth. It would be nice to handle this in a good way (being able to manually disable HEVC direct play to force transcoding?) as it's quite common.
